pilot/pkg/config/memory/monitor.go

```diff
@@ -72,32 +77,45 @@ func newBufferedMonitor(store model.ConfigStore, bufferSize int, sync bool) Moni
 		store:    store,
 		handlers: handlers,
 		eventCh:  make(chan ConfigEvent, bufferSize),
-		sync:     sync,
+		sync:     syncMode,
+		lock:     &sync.RWMutex{},
 	}
 }
 
 func (m *configStoreMonitor) ScheduleProcessEvent(configEvent ConfigEvent) {
+	m.lock.Lock()
+	if m.closed {
+		m.lock.Unlock()
+		return
+	}
+
 	if m.sync {
+		m.lock.Unlock()
 		m.processConfigEvent(configEvent)
 	} else {
 		m.eventCh <- configEvent
+		m.lock.Unlock()
 	}
 }
 
+func (m *configStoreMonitor) run(stop <-chan struct{}) {
+	<-stop
```

Comment:
This function is confusing, actually it is waiting for close notification and do some cleanups

Reply:
what about  `waitAndCleanup`